src/Integration.Vsix.UnitTests/CFamily/CFamilyPreCompiledHeadersEventListenerTests.cs

```diff
@@ -0,0 +1,176 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Utilities;
+using Moq;
+using SonarLint.VisualStudio.Core.Analysis;
+using SonarLint.VisualStudio.Core.CFamily;
+using SonarLint.VisualStudio.Integration.Vsix;
+using SonarLint.VisualStudio.Integration.Vsix.CFamily;
+using SonarLint.VisualStudio.Integration.Vsix.Helpers.DocumentEvents;
+
+namespace SonarLint.VisualStudio.Integration.UnitTests.CFamily
+{
+    [TestClass]
+    public class CFamilyPreCompiledHeadersEventListenerTests
+    {
+        private const string FocusedDocumentFilePath = "c:\\myfile.cpp";
+        private IContentType contentTypeMock;
+
+        private Mock<ICLangAnalyzer> clangAnalyzerMock;
+        private Mock<IActiveDocumentTracker> activeDocumentTrackerMock;
+        private Mock<IScheduler> schedulerMock;
+        private Mock<ISonarLanguageRecognizer> languageRecognizerMock;
+
+        private CFamilyPreCompiledHeadersEventListener testSubject;
+
+        [TestInitialize]
+        public void TestInitialize()
+        {
+            contentTypeMock = Mock.Of<IContentType>();
+            clangAnalyzerMock = new Mock<ICLangAnalyzer>();
+            activeDocumentTrackerMock = new Mock<IActiveDocumentTracker>();
+            schedulerMock = new Mock<IScheduler>();
+            languageRecognizerMock = new Mock<ISonarLanguageRecognizer>();
+
+            testSubject = new CFamilyPreCompiledHeadersEventListener(clangAnalyzerMock.Object, activeDocumentTrackerMock.Object, schedulerMock.Object, languageRecognizerMock.Object);
+        }
+
+        [TestMethod]
+        public void Listen_RegisterToDocumentFocusedEvent()
+        {
+            testSubject.Listen();
+
+            RaiseDocumentFocusedEvent();
+
+            languageRecognizerMock.Verify(x => x.Detect(FocusedDocumentFilePath, contentTypeMock), Times.Once);
+        }
+
+        [TestMethod]
+        public void ListenNotCalled_ShouldNotRegisterToDocumentFocusedEvent()
+        {
+            RaiseDocumentFocusedEvent();
+
+            clangAnalyzerMock.VerifyNoOtherCalls();
+            schedulerMock.VerifyNoOtherCalls();
+            languageRecognizerMock.VerifyNoOtherCalls();
+        }
+
+        [TestMethod]
+        public void Dispose_UnregisterFromDocumentFocusedEvent()
+        {
+            testSubject.Listen();
+            testSubject.Dispose();
+
+            RaiseDocumentFocusedEvent();
+
+            clangAnalyzerMock.VerifyNoOtherCalls();
+            schedulerMock.VerifyNoOtherCalls();
+            languageRecognizerMock.VerifyNoOtherCalls();
+        }
+
+        [TestMethod]
+        public void OnDocumentFocused_NoLanguagesDetected_PchGenerationNotScheduled()
+        {
+            languageRecognizerMock
```

Comment:
This setup step is repeated in three places and only the return value varies - a helper method would be make the tests easier to read.